test/Stripe.hs

```diff
@@ -163,15 +170,43 @@ chargeTests =
       let amount = 650
       let currency = AED
       db <- memory
-      (Left (ServerError code _ _ _)) <- runExceptT . runHandler' $ charge stripeConfig db (Charges token voucher amount currency)
+      (Left (ServerError code phrase body _)) <- runExceptT . runHandler' $ charge stripeConfig db (Charges token voucher amount currency)
       assertEqual "The result is an error" 400 code
+      assertEqual "The HTTP phrase matches the code" "Bad Request" phrase
+      assertEqual "The JSON body includes the reason" (Just $ Failure "Unsupported currency") (decode body)
+
   , testCase "incorrect USD amount is rejected" $
     withFakeStripe (return chargeOkay) $ \stripeConfig -> do
       let amount = 649
       let currency = USD
       db <- memory
-      (Left (ServerError code _ _ _)) <- runExceptT . runHandler' $ charge stripeConfig db (Charges token voucher amount currency)
+      (Left (ServerError code phrase body _)) <- runExceptT . runHandler' $ charge stripeConfig db (Charges token voucher amount currency)
       assertEqual "The result is an error" 400 code
+      assertEqual "The HTTP phrase matches the code" "Bad Request" phrase
+      assertEqual "The JSON body includes the reason" (Just $ Failure "Incorrect charge amount") (decode body)
+
+  , testCase "a Stripe charge failure is propagated" $
+    withFakeStripe (return (chargeFailed cardError)) $ \stripeConfig -> do
+      let amount = 650
+      let currency = USD
+      db <- memory
+      (Left (ServerError code phrase body _)) <- runExceptT . runHandler' $ charge stripeConfig db (Charges token voucher amount currency)
+      assertEqual "The result is an error" 400 code
+      assertEqual "The HTTP phrase matches the code" "Bad Request" phrase
+      -- The `cardError` is for a card expired error.
+      assertEqual "The JSON body includes the reason"
+        (Just $ Failure "Stripe charge didn't succeed: Your card is expired.") (decode body)
```

Comment:
> Your card is expired.

Is this correct?  I thought `cardError` would indicate a whole lot of different things, Above, your own comment says:

>       -- "Card errors are the most common type of error you should expect to
>       -- handle. They result when the user enters a card that can't be charged
>       -- for some reason."


----

Above (Line 186, `let err = errorForStripe errorType ( concat [ "Stripe charge didn't succeed: ", msg ])`) very much looks like the Stripe upstream error message is propagated, which is what we want.
So I read this as "the test fixture `cardError` is a card expired error" and not as "`cardError` will result in a message that says the card is expired".

Reply:
Yep - see https://github.com/PrivateStorageio/PaymentServer/pull/113/files#diff-c44673a5714663692552825bf3666ca8557cf8217b71f349cae6aa1c93e2c10dR53-R62

In the comment I just meant that this specific `cardError` value that we have is for that case.  You could have a different card error for another case (but the test suite doesn't have one).